### PR TITLE
Define mono_class_init_ready and MIN and MAX readiness levels

### PR DESCRIPTION
**Pull request by lambdageek**
# FOR REFERENCE ONLY

Please comment on the [RFC and design discussion](https://github.com/mono/mono/issues/7856)

### CODE REVIEW

File: mono/metadata/class-init.c [hunk 1]
```diff
@@ -4107,6 +4113,17 @@ mono_class_init (MonoClass *klass)
...
 
+	g_assert (readiness >= MONO_CLASS_READY_MIN && readiness <= MONO_CLASS_READY_MAX);
+
+	switch (readiness) {
+	case MONO_CLASS_READY_BAREBONES:
+		return !mono_class_has_failure (klass);
+	case MONO_CLASS_READY_INSTANTIATE:
+		break; /* continue with initialization */
+	default:
+		g_assert_not_reached ();
```
Comment (luhenry) [inline mono/metadata/class-init.c]: a `g_error` with the value passed would help debugging.

**Comment by kumpera** [comment]
LGTM, should this cover the many lazy init steps in the _setup_ functions?

One thing, I believe the comment in MONO_CLASS_READY_INSTANTIATE is incorrect as mono_class_init AFAICT doesn't do vtable setup.

**Comment by lambdageek** [comment]
@kumpera 1. yes this should eventually take over from the `_setup_` functions.  2. thanks - looks like `mono_class_init` only sets up the vtable for generic instances. I'll rename. 3. We have the overall design discussion over at #7856

**Comment by steveisok** [comment]
Since this is older, I'm closing.  Please see the [old pull request policy](https://github.com/mono/mono/wiki/Pull-Request-Guide#old-pull-request-policy) for more info.

**Pull request closed without merging**